File: lua/statuspage_sdk.lua
-- Statuspage SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Typed-model annotations (LuaLS ---@class); empty at runtime.
require("statuspage_types")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local StatuspageSDK = {}
StatuspageSDK.__index = StatuspageSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

StatuspageSDK._make_feature = _make_feature


function StatuspageSDK.new(options)
  local self = setmetatable({}, StatuspageSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

    -- feature: test


  return self
end


function StatuspageSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function StatuspageSDK:get_utility()
  return Utility.copy(self._utility)
end


function StatuspageSDK:get_root_ctx()
  return self._rootctx
end


function StatuspageSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


-- Raw endpoint access is operator-controllable, like every entity op.
-- Blocking it means denying BOTH the 'direct' and 'graphql' tokens, since
-- either one reaches the same endpoint.
function StatuspageSDK:direct(fetchargs)
  if not self:_op_allowed("direct") then
    return self:_op_denied("direct"), nil
  end

  return self:_raw_request(fetchargs)
end


-- Is this raw-access op permitted by the SDK's allow.op option?
function StatuspageSDK:_op_allowed(op)
  local allow = vs.getpath(self.options, "allow.op")
  return type(allow) == "string" and allow:find(op, 1, true) ~= nil
end


function StatuspageSDK:_op_denied(op)
  local allow = vs.getpath(self.options, "allow.op")
  if type(allow) ~= "string" then allow = "" end
  return {
    ok = false,
    err = "StatuspageSDK: " .. op .. ": operation not allowed by" ..
      " SDK option allow.op value: \"" .. allow .. "\"",
  }
end


-- Ungated request path shared by direct and graphql, each of which checks its
-- own allow.op token first. Private, rather than a flag on fetchargs: a
-- caller-supplied marker would let anyone opt straight back out of the gate
-- by passing it.
function StatuspageSDK:_raw_request(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end


-- Raw GraphQL access: the pressure valve that makes the generated surface's
-- deliberate omissions (per-call selection sets, typed filter builders,
-- batching, subscriptions) livable — the whole schema stays reachable.
--
-- Thin wrapper over the same prepare/fetch path direct uses, with the one
-- thing raw direct cannot do for GraphQL: a GraphQL failure rides HTTP 200 as
-- a top-level `errors` array, so status alone would report a failed query as
-- ok.
--
-- NOTE: like direct, this bypasses the feature pipeline — no retry, ratelimit
-- or paging features apply.
function StatuspageSDK:graphql(query, variables, ctrl)
  if not self:_op_allowed("graphql") then
    return self:_op_denied("graphql"), nil
  end

  local res, err = self:_raw_request({
    method = "POST",
    headers = { ["content-type"] = "application/json" },
    body = {
      query = query,
      variables = type(variables) == "table" and variables or {},
    },
    ctrl = type(ctrl) == "table" and ctrl or {},
  })

  if err ~= nil or type(res) ~= "table" then
    return res, err
  end

  -- Errors are read BEFORE any status check: a GraphQL parse or validation
  -- failure comes back as HTTP 400 carrying the standard { errors = {...} }
  -- body, and the raw path represents a non-2xx as ok=false with no err — so
  -- returning early on status would discard the server's own diagnostics,
  -- which are the only useful part of that response.
  local errors = vs.getpath(res, "data.errors")

  if type(errors) == "table" and 0 < #errors then
    local msg = vs.getprop(errors[1], "message")
    if type(msg) ~= "string" or msg == "" then
      msg = "graphql error"
    end
    res.ok = false
    res.err = "StatuspageSDK: graphql: " .. msg
    res.graphql = errors
  end

  return res, nil
end



-- Idiomatic facade: client:Component():list() / client:Component():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:Component(data)
  local EntityMod = require("entity.component_entity")
  if data == nil then
    if self._component == nil then
      self._component = EntityMod.new(self, nil)
    end
    return self._component
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ComponentGroupUptime():list() / client:ComponentGroupUptime():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:ComponentGroupUptime(data)
  local EntityMod = require("entity.component_group_uptime_entity")
  if data == nil then
    if self._component_group_uptime == nil then
      self._component_group_uptime = EntityMod.new(self, nil)
    end
    return self._component_group_uptime
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:GroupComponent():list() / client:GroupComponent():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:GroupComponent(data)
  local EntityMod = require("entity.group_component_entity")
  if data == nil then
    if self._group_component == nil then
      self._group_component = EntityMod.new(self, nil)
    end
    return self._group_component
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Incident():list() / client:Incident():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:Incident(data)
  local EntityMod = require("entity.incident_entity")
  if data == nil then
    if self._incident == nil then
      self._incident = EntityMod.new(self, nil)
    end
    return self._incident
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:IncidentPostmortem():list() / client:IncidentPostmortem():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:IncidentPostmortem(data)
  local EntityMod = require("entity.incident_postmortem_entity")
  if data == nil then
    if self._incident_postmortem == nil then
      self._incident_postmortem = EntityMod.new(self, nil)
    end
    return self._incident_postmortem
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:IncidentSubscriber():list() / client:IncidentSubscriber():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:IncidentSubscriber(data)
  local EntityMod = require("entity.incident_subscriber_entity")
  if data == nil then
    if self._incident_subscriber == nil then
      self._incident_subscriber = EntityMod.new(self, nil)
    end
    return self._incident_subscriber
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:IncidentTemplate():list() / client:IncidentTemplate():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:IncidentTemplate(data)
  local EntityMod = require("entity.incident_template_entity")
  if data == nil then
    if self._incident_template == nil then
      self._incident_template = EntityMod.new(self, nil)
    end
    return self._incident_template
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:IncidentUpdate():list() / client:IncidentUpdate():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:IncidentUpdate(data)
  local EntityMod = require("entity.incident_update_entity")
  if data == nil then
    if self._incident_update == nil then
      self._incident_update = EntityMod.new(self, nil)
    end
    return self._incident_update
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Metric():list() / client:Metric():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:Metric(data)
  local EntityMod = require("entity.metric_entity")
  if data == nil then
    if self._metric == nil then
      self._metric = EntityMod.new(self, nil)
    end
    return self._metric
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:MetricsProvider():list() / client:MetricsProvider():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:MetricsProvider(data)
  local EntityMod = require("entity.metrics_provider_entity")
  if data == nil then
    if self._metrics_provider == nil then
      self._metrics_provider = EntityMod.new(self, nil)
    end
    return self._metrics_provider
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Page():list() / client:Page():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:Page(data)
  local EntityMod = require("entity.page_entity")
  if data == nil then
    if self._page == nil then
      self._page = EntityMod.new(self, nil)
    end
    return self._page
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PageAccessGroup():list() / client:PageAccessGroup():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:PageAccessGroup(data)
  local EntityMod = require("entity.page_access_group_entity")
  if data == nil then
    if self._page_access_group == nil then
      self._page_access_group = EntityMod.new(self, nil)
    end
    return self._page_access_group
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PageAccessUser():list() / client:PageAccessUser():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:PageAccessUser(data)
  local EntityMod = require("entity.page_access_user_entity")
  if data == nil then
    if self._page_access_user == nil then
      self._page_access_user = EntityMod.new(self, nil)
    end
    return self._page_access_user
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Permission():list() / client:Permission():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:Permission(data)
  local EntityMod = require("entity.permission_entity")
  if data == nil then
    if self._permission == nil then
      self._permission = EntityMod.new(self, nil)
    end
    return self._permission
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Postmortem():list() / client:Postmortem():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:Postmortem(data)
  local EntityMod = require("entity.postmortem_entity")
  if data == nil then
    if self._postmortem == nil then
      self._postmortem = EntityMod.new(self, nil)
    end
    return self._postmortem
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:StatusEmbedConfig():list() / client:StatusEmbedConfig():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:StatusEmbedConfig(data)
  local EntityMod = require("entity.status_embed_config_entity")
  if data == nil then
    if self._status_embed_config == nil then
      self._status_embed_config = EntityMod.new(self, nil)
    end
    return self._status_embed_config
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Subscriber():list() / client:Subscriber():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:Subscriber(data)
  local EntityMod = require("entity.subscriber_entity")
  if data == nil then
    if self._subscriber == nil then
      self._subscriber = EntityMod.new(self, nil)
    end
    return self._subscriber
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:User():list() / client:User():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function StatuspageSDK:User(data)
  local EntityMod = require("entity.user_entity")
  if data == nil then
    if self._user == nil then
      self._user = EntityMod.new(self, nil)
    end
    return self._user
  end
  return EntityMod.new(self, data)
end




function StatuspageSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = StatuspageSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return StatuspageSDK

File: js/src/StatuspageSDK.js
// Statuspage Js SDK

const { ComponentEntity } = require('./entity/ComponentEntity')
const { ComponentGroupUptimeEntity } = require('./entity/ComponentGroupUptimeEntity')
const { GroupComponentEntity } = require('./entity/GroupComponentEntity')
const { IncidentEntity } = require('./entity/IncidentEntity')
const { IncidentPostmortemEntity } = require('./entity/IncidentPostmortemEntity')
const { IncidentSubscriberEntity } = require('./entity/IncidentSubscriberEntity')
const { IncidentTemplateEntity } = require('./entity/IncidentTemplateEntity')
const { IncidentUpdateEntity } = require('./entity/IncidentUpdateEntity')
const { MetricEntity } = require('./entity/MetricEntity')
const { MetricsProviderEntity } = require('./entity/MetricsProviderEntity')
const { PageEntity } = require('./entity/PageEntity')
const { PageAccessGroupEntity } = require('./entity/PageAccessGroupEntity')
const { PageAccessUserEntity } = require('./entity/PageAccessUserEntity')
const { PermissionEntity } = require('./entity/PermissionEntity')
const { PostmortemEntity } = require('./entity/PostmortemEntity')
const { StatusEmbedConfigEntity } = require('./entity/StatusEmbedConfigEntity')
const { SubscriberEntity } = require('./entity/SubscriberEntity')
const { UserEntity } = require('./entity/UserEntity')


const { inspect } = require('node:util')

const { config } = require('./Config')
const { Utility } = require('./utility/Utility')
const { StatuspageEntityBase } = require('./StatuspageEntityBase')


const { BaseFeature } = require('./feature/base/BaseFeature')


const stdutil = new Utility()


class StatuspageSDK {
  _mode = 'live'
  _options
  _utility = new Utility()
  _features
  _rootctx

  constructor(options) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    // Add features in the resolved order (makeOptions puts an explicit
    // array order first, else defaults to test-first). Ordering matters:
    // the `test` feature installs the base mock transport and the transport
    // features (retry/cache/netsim/proxy/ratelimit) wrap whatever is current,
    // so `test` must be added before them to sit at the base of the chain.
    const featureorder = getpath(this._options, '__derived__.featureorder') || []
    for (const fname of featureorder) {
      const fopts = this._options.feature[fname] || {}
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    }

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  // Raw endpoint access is operator-controllable, like every entity op.
  // Blocking it means denying BOTH the 'direct' and 'graphql' tokens, since
  // either one reaches the same endpoint.
  async direct(fetchargs) {
    if (!this._options.allow.op.includes('direct')) {
      return {
        ok: false,
        err: new Error('StatuspageSDK: direct: operation not allowed by' +
          ' SDK option allow.op value: "' + this._options.allow.op + '"'),
      }
    }

    return this._rawRequest(fetchargs)
  }


  // Ungated request path shared by direct() and graphql(), each of which
  // checks its own allow.op token first. Private, rather than a flag on
  // fetchargs: a caller-supplied marker would let anyone opt straight back
  // out of the gate by passing it.
  async _rawRequest(fetchargs) {
    const utility = this._utility

    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status
      const json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err) {
      return { ok: false, err }
    }
  }



  // Raw GraphQL access: the pressure valve that makes the generated
  // surface's deliberate omissions (per-call selection sets, typed filter
  // builders, batching, subscriptions) livable — the whole schema stays
  // reachable.
  //
  // Thin wrapper over the same prepare/fetch path `direct` uses, with the
  // one thing raw `direct` cannot do for GraphQL: a GraphQL failure rides
  // HTTP 200 as a top-level `errors` array, so status alone would report a
  // failed query as ok.
  //
  // NOTE: like `direct`, this bypasses the feature pipeline — no retry,
  // ratelimit or paging features apply.
  async graphql(query, variables, ctrl) {
    const options = this._options

    if (!options.allow.op.includes('graphql')) {
      return {
        ok: false,
        err: new Error('StatuspageSDK: graphql: operation not allowed by' +
          ' SDK option allow.op value: "' + options.allow.op + '"'),
      }
    }

    const res = await this._rawRequest({
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: { query, variables: variables || {} },
      ctrl,
    })

    if (res instanceof Error) {
      return res
    }

    // Errors are read BEFORE any status check: a GraphQL parse or validation
    // failure comes back as HTTP 400 carrying the standard { errors: [...] }
    // body, and the raw path represents a non-2xx as { ok: false } with no
    // err — so returning early on status would discard the server's own
    // diagnostics, which are the only useful part of that response.
    const errors = null == res.data ? undefined : res.data.errors

    if (null != errors && Array.isArray(errors) && 0 < errors.length) {
      const first = errors[0] || {}
      const err = new Error('StatuspageSDK: graphql: ' +
        (first.message || 'graphql error'))
      err.graphql = errors
      return { ok: false, status: res.status, headers: res.headers, err, data: res.data }
    }

    return res
  }



  // Entity access: `client.Component().list()` / `client.Component().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Component(entopts) {
    const self = this
    return new ComponentEntity(self, entopts)
  }


  // Entity access: `client.ComponentGroupUptime().list()` / `client.ComponentGroupUptime().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  ComponentGroupUptime(entopts) {
    const self = this
    return new ComponentGroupUptimeEntity(self, entopts)
  }


  // Entity access: `client.GroupComponent().list()` / `client.GroupComponent().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  GroupComponent(entopts) {
    const self = this
    return new GroupComponentEntity(self, entopts)
  }


  // Entity access: `client.Incident().list()` / `client.Incident().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Incident(entopts) {
    const self = this
    return new IncidentEntity(self, entopts)
  }


  // Entity access: `client.IncidentPostmortem().list()` / `client.IncidentPostmortem().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  IncidentPostmortem(entopts) {
    const self = this
    return new IncidentPostmortemEntity(self, entopts)
  }


  // Entity access: `client.IncidentSubscriber().list()` / `client.IncidentSubscriber().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  IncidentSubscriber(entopts) {
    const self = this
    return new IncidentSubscriberEntity(self, entopts)
  }


  // Entity access: `client.IncidentTemplate().list()` / `client.IncidentTemplate().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  IncidentTemplate(entopts) {
    const self = this
    return new IncidentTemplateEntity(self, entopts)
  }


  // Entity access: `client.IncidentUpdate().list()` / `client.IncidentUpdate().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  IncidentUpdate(entopts) {
    const self = this
    return new IncidentUpdateEntity(self, entopts)
  }


  // Entity access: `client.Metric().list()` / `client.Metric().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Metric(entopts) {
    const self = this
    return new MetricEntity(self, entopts)
  }


  // Entity access: `client.MetricsProvider().list()` / `client.MetricsProvider().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  MetricsProvider(entopts) {
    const self = this
    return new MetricsProviderEntity(self, entopts)
  }


  // Entity access: `client.Page().list()` / `client.Page().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Page(entopts) {
    const self = this
    return new PageEntity(self, entopts)
  }


  // Entity access: `client.PageAccessGroup().list()` / `client.PageAccessGroup().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  PageAccessGroup(entopts) {
    const self = this
    return new PageAccessGroupEntity(self, entopts)
  }


  // Entity access: `client.PageAccessUser().list()` / `client.PageAccessUser().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  PageAccessUser(entopts) {
    const self = this
    return new PageAccessUserEntity(self, entopts)
  }


  // Entity access: `client.Permission().list()` / `client.Permission().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Permission(entopts) {
    const self = this
    return new PermissionEntity(self, entopts)
  }


  // Entity access: `client.Postmortem().list()` / `client.Postmortem().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Postmortem(entopts) {
    const self = this
    return new PostmortemEntity(self, entopts)
  }


  // Entity access: `client.StatusEmbedConfig().list()` / `client.StatusEmbedConfig().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  StatusEmbedConfig(entopts) {
    const self = this
    return new StatusEmbedConfigEntity(self, entopts)
  }


  // Entity access: `client.Subscriber().list()` / `client.Subscriber().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Subscriber(entopts) {
    const self = this
    return new SubscriberEntity(self, entopts)
  }


  // Entity access: `client.User().list()` / `client.User().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  User(entopts) {
    const self = this
    return new UserEntity(self, entopts)
  }




  static test(testoptsarg, sdkoptsarg) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new StatuspageSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts, sdkopts) {
    return StatuspageSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'Statuspage' }
  }

  toString() {
    return 'Statuspage ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = StatuspageSDK


module.exports = {
  stdutil,
  config,

  BaseFeature,
  StatuspageEntityBase,

  StatuspageSDK,
  SDK,
}

